### ISSUE TITLE
Better UI for annotate feature

### ISSUE THREAD

**Issue by OpenNingia** [OpenNingia/vscode-clearcase#12]
I added support for the `cleartool annotate` command ( in some situations it's a godsent ) but the UI is terrible ( it just open a text document with the output of the cleartool command ).

We need a better way to display those informations ( for example https://github.com/eamodio/vscode-gitlens does a **great** work! )

**Comment by fr43nk** [comment]
I never used it before but looks like an interesting feature :)
At first we could try to create a two column view into the currently open file and avoid open a second editor window. Right now I have no idea how to do that, but maybe we can find some information about that somewhere.

**Comment by fr43nk** [comment]
@OpenNingia A small update: It is still wip, but I am able to display the annotation right in the currently open file. I think, tomorrow I can provide a branch for you, so you can have a look at it. It took me some time to understand how this works...

**Comment by fr43nk** [comment]
Sorry, but I because of my sources I cannot provide a screenshot.
But if you want, you can checkout my branch to yourself :)

https://github.com/fr43nk/vscode-clearcase/tree/feature/cc-codelens-01

It is still wip.

**Comment by OpenNingia** [comment]
Nice, thanks. 
Here some feedback:

- Toggling Annotation is not intuitive ( you need to switch file, then come back )
- Annotation text removes '\' character, you should `.replace(/\\/g, "/").trim()` 
- Inline annotation doesn't play nice with checked-out files, I suggest to remove the `-nco` parameter from the cleartool command and display an error when annotating checked-out files

Wishlist / near future:
- When annotation is enabled code folding become very chaotic, is there a way to disable folding?
- Customizable format maybe?
- Different color for date/time, user and branch

**Comment by fr43nk** [comment]
I found some time to do some coding on the annotation feature.
Things I addressed:
- renamed the the code lens (toggle should work without changing the editor)
- .replace command for annotation section
- error message if file is checked out
- added configuration of the color and background

Still todo:
- code folding detection: switch it off if enabled
- different colors of user and date (right now I have no idea how to do this, there is only one style for the whole text that is inserted before the existing one in the line

What would be nice to have:
- a kind of a diff view if the annotation is used on a checked out file instead of an error message

The branch is the same as in my previous post.

**Comment by OpenNingia** [comment]
Wow, now is very nice. I've made an addition that you might incorporate in your code. I added a new parameter `AnnotationFormatString` to customize the annotation string.

```
public async getAnnotatedFileContent(filePath: string): Promise<string> {
        let fmt = this.configHandler.configuration.AnnotationFormatString;
        let sep = " | ";
        let param = "\"" + filePath + "\"";
        let cmd = "cleartool annotate -out - -nhe -fmt \"" + fmt + sep + "\" " + param;
```

The formatting option are documented [here](https://www.ibm.com/support/knowledgecenter/SSSH27_8.0.0/com.ibm.rational.clearcase.cc_ref.doc/topics/fmt_ccase.htm)

The default is something like: `private m_annotationFormatString: string = "%d %12u";`

**Comment by fr43nk** [comment]
Thanks.
Great idea adding this config. I did it as you have proposed and it seems to work.

What do you think, should we do anything additional about this annotation feature or is it now ready for production?

**Comment by OpenNingia** [comment]
I think it's ok to ship it :+1: 
Go on and create a pull request when you're ready.

**Comment by OpenNingia** [comment]
Released in 1.9.0, thanks :1st_place_medal:
